source/particle/interpolator/bilinear_least_squares.cc

```diff
@@ -0,0 +1,154 @@
+                                                           this->get_triangulation(),
+                                                           approximated_cell_midpoint)).first;
+          }
+        else
+          found_cell = cell;
+
+        const types::LevelInd cell_index = std::make_pair<unsigned int, unsigned int> (found_cell->level(),found_cell->index());
+        const std::pair<typename std::multimap<types::LevelInd, Particle<dim> >::const_iterator,
+              typename std::multimap<types::LevelInd, Particle<dim> >::const_iterator> particle_range = particles.equal_range(cell_index);
+
+        const unsigned int n_particles = std::distance(particle_range.first,particle_range.second);
+        const unsigned int n_particle_properties = particles.begin()->second.get_properties().size();
+
+        std::vector<std::vector<double> > cell_properties(positions.size(),
+                                                          std::vector<double>(n_particle_properties, numbers::signaling_nan<double>()));
+
+        unsigned int property_index = 0;
+        for (unsigned int i=0; i < n_particle_properties; i++)
+          if (selected_properties[i])
+            property_index = i;
+
+        AssertThrow(n_particles != 0,
+                    ExcMessage("At least one cell contained no particles. The 'bilinear'"
+                               "interpolation scheme does not support this case. "));
+
+
+        const unsigned int matrix_dimension = 4;
+        dealii::LAPACKFullMatrix<double> A(n_particles, matrix_dimension);
+        Vector<double> r(n_particles);
+        A = 0;
+        r = 0;
+
+        unsigned int index = 0;
+        double cell_diameter = found_cell->diameter();
+        for (typename std::multimap<types::LevelInd, Particle<dim> >::const_iterator particle = particle_range.first;
+             particle != particle_range.second; ++particle, ++index)
+          {
+            const double particle_property_value = particle->second.get_properties()[property_index];
+            r[index] = particle_property_value;
+
+            const Point<dim> position = particle->second.get_location();
+            A(index,0) = 1;
+            A(index,1) = (position[0] - approximated_cell_midpoint[0])/cell_diameter;
+            A(index,2) = (position[1] - approximated_cell_midpoint[1])/cell_diameter;
+            A(index,3) = (position[0] - approximated_cell_midpoint[0]) * (position[1] - approximated_cell_midpoint[1])/std::pow(cell_diameter,2);
+          }
+
+        dealii::LAPACKFullMatrix<double> B(matrix_dimension, matrix_dimension);
+        dealii::LAPACKFullMatrix<double> B_inverse(B);
+
+        Vector<double> c_ATr(matrix_dimension);
+        Vector<double> c(matrix_dimension);
+
+        const double threshold = 1e-15;
+        unsigned int index_positions = 0;
+
+        A.Tmmult(B, A, false);
+        A.Tvmult(c_ATr,r);
+        B_inverse.compute_svd();
+        B_inverse.compute_inverse_svd(threshold);
```

Comment:
Do you need both of the calls above? I am not expert in Least squares, and even less in SVD, so please check both calls are necessary and add a comment that explains what is happening in these four lines.